auth/user_mgt.go

```diff
@@ -561,6 +582,261 @@ func (c *baseClient) getUser(ctx context.Context, query *userQuery) (*UserRecord
 	return parsed.Users[0].makeUserRecord()
 }
 
+// A UserIdentifier identifies a user to be looked up.
+type UserIdentifier interface {
+	String() string
+
+	validate() error
+	matchesUserRecord(ur *UserRecord) (bool, error)
+	populateRequest(req *getAccountInfoRequest) error
+}
+
+// A UIDIdentifier is used for looking up an account by uid.
+//
+// See GetUsers function.
+type UIDIdentifier struct {
+	UID string
+}
+
+func (id UIDIdentifier) String() string {
+	return fmt.Sprintf("UIDIdentifier{%s}", id.UID)
+}
+
+func (id UIDIdentifier) validate() error {
+	return validateUID(id.UID)
+}
+
+func (id UIDIdentifier) matchesUserRecord(ur *UserRecord) (bool, error) {
```

Comment:
We can move this logic to `getAccountInfoRequest`. In pseudo code:

```
req := InitGetAccountInfoRequest()
Populate(req, uids)
result := MakeRPC(req)
for _, uid := range req.localId {
  // Search for missing UIDs
}
for _, email := range req.email {
  // Search for missing emails
}
// And so on for phone and federatedIds
```

This looks a bit crass, but it does have a couple of advantages:
1. We don't have to validate the identifiers after the RPC has been made. Since we are not looking at the user arguments once the request has been created, we know it hasn't been modified by the user.
2. We can remove `validate()` and `matchesUserRecord()` from this interface.

Reply:
(The validate() thing is no longer applicable.)

I'm not sure this is a worthwhile trade-off. Without the extra interface method (matchesUserRecord) we have to test all possible "sub-types". If we add a new subtype, we won't have any compile time safety to ensure the check is added here.

Additionally, we'll have to check to see if each type is either a (eg) UIDIdentifier or a *UIDIdentifier. I don't know a way to do that without duplicating the code. (But I don't know go very well, so suggestions are welcome.)

I've gone ahead and done this anyways, but in a separate commit. I'd prefer to just drop the commit, but if you prefer it (possibly with some improvements), then we can include it.

If we must get down to <=2 methods in the interface, I'd prefer to drop toString(). It's not currently used anywhere, and since it's not private, it can't be used by the end user either. (In fact, I think I'd like to drop it regardless of this issue; see my other comment.)